ccl/changefeedccl: TestChangefeedNemeses failed

[(ccl/changefeedccl).TestChangefeedNemeses failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1922696&tab=buildLog) on [master@19ab7176c362177c03756424136a4c8d056be3b4](https://github.com/cockroachdb/cockroach/commits/19ab7176c362177c03756424136a4c8d056be3b4):

```
=== RUN   TestChangefeedNemeses
--- FAIL: TestChangefeedNemeses (14.61s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestChangefeedNemeses137369527
    test_log_scope.go:58: use -show-logs to present logs inline
=== RUN   TestChangefeedNemeses/enterprise
    --- FAIL: TestChangefeedNemeses/enterprise (5.11s)
        nemeses_test.go:38: could not pause job
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest.(*jobFeed).Pause.func1
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest/testfeed.go:282
            github.com/cockroachdb/cockroach/pkg/util/retry.WithMaxAttempts
            	/go/src/github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:159
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest.(*jobFeed).Pause
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest/testfeed.go:276
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest.pause
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest/nemeses.go:654
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest.logEvent.func1
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest/nemeses.go:499
            github.com/cockroachdb/cockroach/pkg/util/fsm.Transitions.apply
            	/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:106
            github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).ApplyWithPayload
            	/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:135
            github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).Apply
            	/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:129
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest.RunNemesis
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest/nemeses.go:215
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedNemeses.func2
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/nemeses_test.go:36
            github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.enterpriseTestWithServerArgs.func1
            	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:288
            testing.tRunner
            	/usr/local/go/src/testing/testing.go:909
            runtime.goexit
            	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestChangefeedNemeses PKG=./pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #47153 ccl/changefeedccl: TestChangefeedNemeses failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1)

- #46706 ccl/changefeedccl: TestChangefeedNemeses failed under stress  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestChangefeedNemeses.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/changefeedccl).TestChangefeedNemeses failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1931062&tab=buildLog) on [master@1d6db7e386e3ab21bca53f38637014f61153dc90](https://github.com/cockroachdb/cockroach/commits/1d6db7e386e3ab21bca53f38637014f61153dc90):

```
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/schemafeed/schema_feed.go:180 +0xd5
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/schemafeed.(*SchemaFeed).Run-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/schemafeed/schema_feed.go:156 +0x55
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x4e
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x71

Previous read at 0x00c006983690 by goroutine 147:
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.copyFromSourceToSinkUntilTableEvent.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:329 +0x160
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.copyFromSourceToSinkUntilTableEvent()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:379 +0x37f
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.(*kvFeed).runUntilTableEvent.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:268 +0x1ba
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x4e
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x71

Goroutine 379 (running) created at:
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x73
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:165 +0xda
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:72 +0x608
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).startKVFeed.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:229 +0x15d
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:323 +0x162

Goroutine 147 (running) created at:
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x73
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:165 +0xda
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.(*kvFeed).runUntilTableEvent()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:267 +0x367
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.(*kvFeed).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:181 +0x12b
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.(*kvFeed).run-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed/kv_feed.go:172 +0x55
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x4e
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x71
==================
    --- FAIL: TestChangefeedNemeses/enterprise (92.27s)
        testing.go:853: race detected during execution of test

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestChangefeedNemeses PKG=./pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #47153 ccl/changefeedccl: TestChangefeedNemeses failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1)

- #46706 ccl/changefeedccl: TestChangefeedNemeses failed under stress  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestChangefeedNemeses.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
+++ b/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
@@ -418,12 +418,12 @@ func (tf *SchemaFeed) ingestDescriptors(
 	ctx context.Context,
 	startTS, endTS hlc.Timestamp,
 	descs []catalog.Descriptor,
-	validateFn func(ctx context.Context, desc catalog.Descriptor) error,
+	validateFn func(ctx context.Context, earliestTsBeingIngested hlc.Timestamp, desc catalog.Descriptor) error,
 ) error {
 	sort.Slice(descs, func(i, j int) bool { return descLess(descs[i], descs[j]) })
 	var validateErr error
 	for _, desc := range descs {
-		if err := validateFn(ctx, desc); validateErr == nil {
+		if err := validateFn(ctx, startTS, desc); validateErr == nil {
 			validateErr = err
 		}
 	}
@@ -482,7 +482,9 @@ func formatEvent(e TableEvent) string {
 	return fmt.Sprintf("%v->%v", formatDesc(e.Before), formatDesc(e.After))
 }
 
-func (tf *SchemaFeed) validateDescriptor(ctx context.Context, desc catalog.Descriptor) error {
+func (tf *SchemaFeed) validateDescriptor(
+	ctx context.Context, earliestTsBeingIngested hlc.Timestamp, desc catalog.Descriptor,
+) error {
 	tf.mu.Lock()
 	defer tf.mu.Unlock()
 	switch desc := desc.(type) {
@@ -528,9 +530,16 @@ func (tf *SchemaFeed) validateDescriptor(ctx context.Context, desc catalog.Descr
 				return err
 			}
 			if !shouldFilter {
+				// Only sort the tail of the events from earliestTsBeingIngested.
+				// The head could already have been handed out and sorting is not
+				// stable.
+				idxToSort := sort.Search(len(tf.mu.events), func(i int) bool {
+					return !tf.mu.events[i].After.ModificationTime.Less(earliestTsBeingIngested)
+				})
 				tf.mu.events = append(tf.mu.events, e)
-				sort.Slice(tf.mu.events, func(i, j int) bool {
-					return descLess(tf.mu.events[i].After, tf.mu.events[j].After)
+				toSort := tf.mu.events[idxToSort:]
+				sort.Slice(toSort, func(i, j int) bool {
+					return descLess(toSort[i].After, toSort[j].After)
 				})
 			}
 		}

--- a/pkg/ccl/changefeedccl/schemafeed/schema_feed_test.go
+++ b/pkg/ccl/changefeedccl/schemafeed/schema_feed_test.go
@@ -26,7 +26,7 @@ func TestTableHistoryIngestionTracking(t *testing.T) {
 
 	ctx := context.Background()
 	ts := func(wt int64) hlc.Timestamp { return hlc.Timestamp{WallTime: wt} }
-	validateFn := func(_ context.Context, desc catalog.Descriptor) error {
+	validateFn := func(_ context.Context, _ hlc.Timestamp, desc catalog.Descriptor) error {
 		if desc.GetName() != `` {
 			return errors.Newf("descriptor: %s", desc.GetName())
 		}

